Cameron/gate nav

newtoncn opened the pull request.

newtoncn commented:
Yeah I guess this is a bit bigger than I expected with the changes to the simulator. I dunno where those extra files came from, something's up there. I would hate to delete the commented code from the test files, it was really a visual test and the comments were the ways to visualize it, I guess I can put them in a util file or something and just not use them for now.

reoliveira commented:
I understand the sentiment of not wanting to delete code you wrote, but I'll try to convince you of why it's a good idea.

Firstly, one of the nice things about having version control is that if you really feel like you need that code again and can't remember how to do it, you can always go and find it again because it's stored permanently in git. But, if code isn't being used actively by either the functionality or an automated test, it should just be deleted. This is for a few reasons: keeping the repo "clean", avoiding confusing the reader, avoiding staleness. Imagine you have
```
doThingA()
// commented
// code
doThingB()
// commented
// code
// more
// comments
doThingC()
```
A couple common things that can happen here is that either the reader misses doThingB() entirely and is unaware anything was done, or the commented code isn't uncommented until a long time later because it is always ignored and then finally needed, only to find that all the functions it was calling are gone, incompatible with the way the function currently runs, or deprecated. And so if it isn't important enough to not be commented out, it isn't important enough to be kept.

In this case if the code is just a way for a developer running the test to visualise what is happening, well 99.9% of the time that test is run it won't be supervised, so the automation should do that check for you. And then if the test starts failing and the developer wants to add in a few lines to pop open the images to see what's going on then it's fairly trivial work to do so, so it can be done when necessary.

Hope this helps rationalise the position of keeping commented code out of the repo.

reoliveira commented:
I would be okay with that. Probably makes sense to add it to src/vision/test/ImageTestUtils.cpp

reoliveira commented on src/gate_detect/include/Gate.h:
```diff
@@ -0,0 +1,117 @@
+
+    void horLineCheck();
+
+  private:
```
picky, but please review naming guidelines, these variables are all named incorrectly and also inconsistently. Also please consider if a variable needs to be a permanent member or if they can be temporary variables.

newtoncn replied:
Changed many variables to be local instead of global

reoliveira commented on src/gate_detect/include/Gate.h:
```diff
@@ -0,0 +1,117 @@
+    cv::Mat dst;
+    int lowVertThresh, lowHorThresh, cannyLow, cannyHigh, poleMax,
+    camera_res_width;
+    float m, b;
```
needs better names, what are these?

newtoncn replied:
Entered detailed descriptions for HoughLinesP and Canny filter

reoliveira commented on src/gate_detect/include/Gate.h:
```diff
@@ -0,0 +1,117 @@
+    void horLineCheck();
+
+  private:
+    cv::Mat dst;
```
what is dst?

newtoncn replied:
Added descriptive names for matrices

reoliveira commented on src/gate_detect/src/Gate.cpp:
```diff
@@ -0,0 +1,189 @@
+
+    if (lines.empty()) return std::vector<float>(9, 0);
+
+    _vertLines = filterVertLines(lines);
```
can any of the variables used here be temporary variables rather than members?

newtoncn replied:
Made lots of vars temporary

reoliveira commented on src/gate_detect/src/Gate.cpp:
```diff
@@ -0,0 +1,189 @@
+
+std::vector<int> Gate::filterVertLines(std::vector<cv::Vec4i> allLines) {
+    std::vector<int> vertLines;
+    vertLines.clear();
```
shouldn't need to clear this vector as it was just created.

newtoncn replied:
Removed

reoliveira commented on src/gate_detect/src/Gate.cpp:
```diff
@@ -0,0 +1,189 @@
+
+std::vector<int> Gate::filterHorLines(std::vector<cv::Vec4i> allLines) {
+    std::vector<int> horLines;
+    horLines.clear();
```
shouldn't need to clear this vector as it was just created.

newtoncn replied:
Removed

reoliveira commented on src/gate_detect/src/Gate.cpp:
```diff
@@ -0,0 +1,189 @@
+std::vector<std::vector<float>>
+Gate::findVertPoles(std::vector<int> vertLines) {
+    std::vector<std::vector<float>> vertPoles;
+    vertPoles.clear();
```
just created this vector

newtoncn replied:
Changed

reoliveira commented on src/gate_detect/src/Gate.cpp:
```diff
@@ -0,0 +1,189 @@
+
+std::vector<std::vector<float>> Gate::findHorPoles(std::vector<int> horLines) {
+    std::vector<std::vector<float>> horPoles;
+    horPoles.clear();
```
just created this vector.

newtoncn replied:
Changed scheme

reoliveira commented on src/gate_detect/src/GateDetectNode.cpp:
```diff
@@ -0,0 +1,73 @@
+void GateDetectNode::publishGateDetectMsg(std::vector<float> gateVectorIn) {
+    gate_detect::gateDetectMsg msg;
+
+    if (gateVectorIn[0] != 0.0)
```
please encapsulate all branch statements with braces.

reoliveira commented on src/gate_detect/test/Testutils.cpp:
```diff
@@ -0,0 +1,35 @@
+#include "TestUtils.h"
```
File name should be `TestUtils.cpp`

garethellis0 commented on src/gate_detect/src/Gate.cpp:
```diff
@@ -0,0 +1,189 @@
+    cannyHigh = 200; // Sets the high threshold for the canny filter, adjust if
+                     // dst not displaying all lines
+    poleMax =
+    200; // Sets the minimum threshold for two lines to be called a pole
```
I think the culprit for the odd formatting here is line 23 of the `.clang-format` file: `ContinuationIndentWidth: '0'`. Feel free to delete it, I'm not a huge fan, removing it will probably make things look nicer.

newtoncn replied:
I put all descriptions in the header files now so this shouldn't be an issue

garethellis0 commented:
Looks like Reid got a bunch of stuff here, in particular to do with naming, so I'll hold off reviewing until the current set of comments are sorted out to avoid duplicate work. Feel free to ping me once you have the current set of comments squared away!

newtoncn commented on src/gate_detect/include/Gate.h:
```diff
@@ -0,0 +1,117 @@
+//
```
Changed naming scheme of node to be more representative of function

newtoncn commented on src/gate_detect/msg/gateDetectMsg.msg:
```diff
@@ -1,9 +1,9 @@
 bool detectLeft
```
Done

newtoncn commented on src/gate_detect/src/Gate.cpp:
```diff
@@ -0,0 +1,189 @@
+#include "Gate.h"
```
Naming changed

newtoncn commented on src/gate_detect/src/gateDetect.cpp:
```diff
@@ -0,0 +1,13 @@
+//
```
Done

newtoncn commented on src/gate_detect/test/gateDetectTest.cpp:
```diff
@@ -0,0 +1,135 @@
+//
```
Done

newtoncn commented on src/gate_detect/test/gateDetectTest.cpp:
```diff
@@ -0,0 +1,135 @@
+// Created by da-cam on 19/05/18.
+//
+
+// Bring in my package's API, which is what I'm testing
```
Removed comment

newtoncn commented on src/gate_detect/test/gateDetectTest.cpp:
```diff
@@ -0,0 +1,135 @@
+using namespace cv;
+using namespace std;
+
+// Declare a test
```
Removed comment

newtoncn commented on src/gate_detect/test/gateDetectTest.cpp:
```diff
@@ -0,0 +1,135 @@
+    EXPECT_TRUE(true);
+}
+
+// Declare a test
```
Removed

newtoncn commented on src/gate_detect/test/gateDetectTest.cpp:
```diff
@@ -0,0 +1,135 @@
+    EXPECT_TRUE(true);
+}
+
+// Declare a test
```
Removed

newtoncn commented on src/gate_detect/test/gateDetectTest.cpp:
```diff
@@ -0,0 +1,135 @@
+    EXPECT_TRUE(true);
+}
+
+// Declare a test
```
Removed

newtoncn commented on src/gate_detect/test/gateDetectTest.cpp:
```diff
@@ -0,0 +1,135 @@
+    EXPECT_TRUE(true);
+}
+
+// Declare a test
```
Removed

newtoncn commented on src/gate_detect/test/gateDetectTest.cpp:
```diff
@@ -0,0 +1,135 @@
+    EXPECT_TRUE(true);
+}
+
+// Declare a test
```
Removed

newtoncn commented on src/gate_detect/test/gateDetectTest.cpp:
```diff
@@ -0,0 +1,135 @@
+int main(int argc, char** argv) {
+    testing::InitGoogleTest(&argc, argv);
+
+    return RUN_ALL_TESTS();
```
Removed

newtoncn commented on src/gate_detect/include/Gate.h:
```diff
@@ -0,0 +1,117 @@
+     * Function to be called on cv::Mat to find vertical poles and horizontal
+     * poles
+     *
+     * mat : cv::Mat potentially containing a gate
```
Removed all spaces. Lots of hugging going on now, no code feels left out.

newtoncn commented on src/gate_detect/include/Gate.h:
```diff
@@ -0,0 +1,117 @@
+     * mat : cv::Mat potentially containing a gate
+     *
+     */
+
```
Removed all spaces. Lots of hugging going on now, no code feels left out

newtoncn commented on src/gate_detect/include/Gate.h:
```diff
@@ -0,0 +1,117 @@
+     *
+     * yf = end coordinate of end of line
+     *
+     */
```
Removed all spaces. Lots of hugging going on now, no code feels left out

newtoncn commented on src/gate_detect/include/Gate.h:
```diff
@@ -0,0 +1,117 @@
+     * yf = end coordinate of end of line
+     *
+     */
+
```
Removed all spaces. Lots of hugging going on now, no code feels left out

newtoncn commented on src/gate_detect/include/Gate.h:
```diff
@@ -0,0 +1,117 @@
+    *
+    */
+
+    std::vector<std::vector<float>> findHorPoles(std::vector<int> horLines);
```
Removed all spaces. Lots of hugging going on now, no code feels left out

newtoncn commented on src/gate_detect/include/Gate.h:
```diff
@@ -0,0 +1,117 @@
+     *
+     */
+
+    std::vector<std::vector<float>> findVertPoles(std::vector<int> vertLines);
```
Removed all spaces. Lots of hugging going on now, no code feels left out

newtoncn commented on src/gate_detect/include/Gate.h:
```diff
@@ -0,0 +1,117 @@
+     * will call a pole
+     *
+     */
+    std::vector<float> gateVector(std::vector<std::vector<float>> vertPoles,
```
Changed to javadoc

newtoncn commented:
After doing a lot of testing I changed the overall structure of this node to preserve more information that improved the capabilities of the node. I've taken note of the formatting comments and fixed the formating of the code. I've also added more descriptions to the header files for reference, let me know if it's too much. I just think it's handy to have in the code.  This node will probably be changed many times over, but I think this new structure is much more flexible.  I've also made some changes where necessary to hsv-adjusting.

newtoncn commented:
@garethellis0 @reoliveira I think this PR should be up to standard now, sorry for all the changes and updates

newtoncn commented:
There's a lot of documentation in GateDetection.h, maybe too much

garethellis0 commented on src/gate_detect/src/GateDetection.cpp:
```diff
@@ -0,0 +1,330 @@
+                line2++;
+            }
+        }
+    }
```
We should add a bunch of comments to this function explaining what we're doing at each step. It would also be helpful if for complex conditionals we split them out. For example `dx_start = abs((*line1)[0] - (*line2)[0])`

newtoncn replied:
defined comments for algorithm

garethellis0 commented on src/gate_detect/src/GateDetection.cpp:
```diff
@@ -0,0 +1,330 @@
+        }
+    }
+
+    return horizontalPoles;
```
Similar comment to above - we should add some comments to explain what we're doing here and break out complex conditionals into variables.

newtoncn replied:
defined comments for algorithm

garethellis0 commented on src/gate_detect/src/GateDetection.cpp:
```diff
@@ -0,0 +1,330 @@
+std::vector<float>
+GateDetection::getGateVector(std::vector<PoleDetection> vertPoles,
+                             std::vector<PoleDetection> horPoles) {
+    std::vector<float> gateCoord(9, 0.0f);
```
I think we should definitly make a `GateCoordinates` struct instead of just using a vector of floats, `gate_coordinates.lower_left_corner_x` is a lot easier to read then `gateCoordinates[3]`. This also helps prevent issues like putting things in the vector in the wrong order, or making the vector too long or short.

newtoncn replied:
Made gateVector into a struct

garethellis0 commented on src/gate_detect/src/GateDetection.cpp:
```diff
@@ -0,0 +1,330 @@
+            gateCoord[5] = rightPole.getVertDistance();
+        }
+    }
+    return gateCoord;
```
This function is very long and seems somewhat repetitive. Is there any way we can shorten it by making a few functions?

garethellis0 commented on src/gate_detect/src/GateDetectionNode.cpp:
```diff
@@ -0,0 +1,114 @@
+    publishGateDetectMsg(gateVector);
+
+    // Uncomment if you want to view gate seen by node
+    // publishGateImage(gateVector);
```
We should make this conditional on a Rosparam, ie:
```
if (view_gate){
    publishGateImage(gateVector)
}
```
(where `view_gate` is set by a rosParam) 

This makes it really easy to turn it on/off without having to recompile the code, and also removes commented out code.

garethellis0 commented on src/gate_detect/src/PoleDetection.cpp:
```diff
@@ -0,0 +1,62 @@
+    _m = m;
+
+    _b = b;
+}
```
See above comments on using initializer lists for constructors for both constructors here.

garethellis0 commented on src/gate_detect/src/PoleDetection.cpp:
```diff
@@ -0,0 +1,62 @@
+float PoleDetection::getVertAngle(int pixelWidthOfImage) {
+    return asin(((((pixelWidthOfImage / 2) - getVertMid()) * (.3048 / 4)) /
+                 getVertWidth()) /
+                getVertDistance());
```
We should break this out into several statements, having this as one long one is kinda confusing

garethellis0 replied:
What is `.3048 / 4`? If it's something we might want to change it should be a rosparam, if it's a constant, then make a `static const int` class member (so that it's named and easily changed)

garethellis0 commented on src/gate_detect/src/PoleDetection.cpp:
```diff
@@ -0,0 +1,62 @@
+float PoleDetection::getHorAngle(int pixelHeightOfImage) {
+    return asin(
+    ((((pixelHeightOfImage / 2) - getHorMid()) * (.3048 / 4)) / getHorWidth()) /
+    getHorDistance());
```
Similar comments to above about breaking this up and making the numbers rosparams or class constants.

garethellis0 commented on src/gate_detect/src/gate_detect_node.cpp:
```diff
@@ -0,0 +1,17 @@
+
+    GateDetectionNode gateDetect = GateDetectionNode(argc, argv);
+
+    return 0;
```
I think you probably need call `ros::spin()` before you return, otherwise this will just immediately exit?

garethellis0 commented on src/gate_detect/test/gate-detection-test.cpp:
```diff
@@ -0,0 +1,134 @@
+    {
+        cout << "Could not open or find the image" << std::endl;
+        FAIL();
+    }
```
We seem to get an image and check if it exists a lot. Might be helpful to add a function that does this to the top of this file to avoid repetitive code.

garethellis0 commented on src/gate_detect/test/gate-detection-test.cpp:
```diff
@@ -0,0 +1,134 @@
+    std::vector<float> floater = gateDetection.initialize(image);
+
+    // Uncomment to view gate drawn over image
+    // TestUtils::DisplayGateDetected(image,floater);
```
We should probably put these display statements inside a conditional and just make it constant at the top ie:
```
if (DRAW_DEBUG_IMAGES){
    TestUtils::DisplayGateDetected(image,floater);
}
````

garethellis0 replied:
We don't actually seem to be checking anything in this test? Same goes for a lot of the tests here?

newtoncn replied:
They're really visual tests, if you uncomment the TestUtils you can see the results. I don't really know how to write assert_true_if tests for this kind of functionality. I was going to do a more in depth test suite but it seems like a time sink

garethellis0 commented on src/gate_detect/test/testUtilities/TestUtils.cpp:
```diff
@@ -0,0 +1,92 @@
+
+    topPole = (int) ((1056 / 2) - ((sin(floatVec[8]) * floatVec[7] * 4.0 *
+                                    ((floatVec[7] - 5.134) / -0.002673)) /
+                                   (.3048)));
```
There a lot of magic numbers here? We should make them constants so that it's more clear what we're doing. A few comments would also be helpful.

newtoncn replied:
A lot of these numbers depend on the interpolation function we're using to define the distance and angle.

reoliveira replied:
That's good information to put in a comment:

`we have chosen to use this interpolation function which relies on these named values ...`

And then have all the dependent numbers as named constants so that readers have at least an idea of what's going on.

garethellis0 commented on src/gate_detect/test/testUtilities/TestUtils.cpp:
```diff
@@ -0,0 +1,92 @@
+
+    topPole = (int) ((imgHeight / 2) - ((sin(floatVec[8]) * floatVec[7] * 4.0 *
+                                         ((floatVec[7] - 5.134) / -0.002673)) /
+                                        (.3048)));
```
See comment above about magic numbers constants and adding comments.

newtoncn replied:
Defined constants, may not have caught all of them but if I find any more in the next couple weeks I'll include them in PR

garethellis0 commented on src/gate_detect/test/testUtilities/TestUtils.h:
```diff
@@ -0,0 +1,20 @@
+//
+// Created by da-cam on 15/06/18.
+//
```
Should have standard start of file comment.

newtoncn replied:
Added

newtoncn replied:
Done

garethellis0 commented on src/simulator/scenes/gate/gateFront.xml:
```diff
@@ -293,43 +299,216 @@
     </offsetp>
     <offsetr>
-      <x> 3.1415</x>
+      <x> 3.14</x>
```
Why did we drop precision here?

newtoncn replied:
.001 radians is .05 degrees, I just stuck to the precision they used initially

reoliveira commented:
Will look more at this later, but as a quick comment:

I just checked and apparently this missed our README, but class data members should be `data_member_` not `_data_member`. Trailing underscore not leading. Will add to README.

newtoncn commented on src/gate_detect/include/GateDetection.h:
```diff
@@ -0,0 +1,181 @@
+/*
+ * Created By: Cameron Newton
+ * Created On: February 27th, 2018
+ */
```
Done

newtoncn commented on src/gate_detect/include/GateDetection.h:
```diff
@@ -0,0 +1,181 @@
+     * poles
+     *
+     * mat : cv::Mat potentially containing a gate
+     */
```
So did you want me to change this?

reoliveira replied:
yes please, all function documentation should follow those guidelines

newtoncn commented on src/gate_detect/include/GateDetection.h:
```diff
@@ -0,0 +1,181 @@
+     *
+     * mat : cv::Mat potentially containing a gate
+     */
+    std::vector<float> initialize(const cv::Mat matin);
```
The definition of the gate vector is in the getGateVector function, I could move it here though

newtoncn commented on src/gate_detect/include/GateDetection.h:
```diff
@@ -0,0 +1,181 @@
+    std::vector<float> initialize(const cv::Mat matin);
+
+  private:
+    int _lowVertThresh, _lowHorThresh;
```
Added def

newtoncn commented on src/gate_detect/include/GateDetection.h:
```diff
@@ -0,0 +1,181 @@
+    * (following Canny's recommendation)
+    * kernel_size: We defined it to be 3 (the size of the Sobel kernel to be
+    * used internally)
+    */
```
I put in the definitions of the Canny and hough lines functions just for ease of use. I could also just link to the definitions online though

newtoncn commented on src/gate_detect/include/GateDetection.h:
```diff
@@ -0,0 +1,181 @@
+     * between pixel width and distance
+     * gathered from calibration
+     */
+    float m_Hor, m_Vert, b_Hor, b_Vert;
```
done

newtoncn commented on src/gate_detect/include/GateDetection.h:
```diff
@@ -0,0 +1,181 @@
+     *
+     *          yf = end coordinate of end of line
+     *
+     * @return  Vector of all cv::Vec4i lines that are horizontal
```
Done

newtoncn commented on src/gate_detect/include/GateDetection.h:
```diff
@@ -0,0 +1,181 @@
+     *
+     *          xf = end coordinate of end of line
+     *
+     *          yf = end coordinate of end of line
```
Done

newtoncn commented on src/gate_detect/include/GateDetection.h:
```diff
@@ -0,0 +1,181 @@
+     *
+     *          yf = end coordinate of end of line
+     *
+     *  @return Vector of all cv::Vec4i lines that are vertical
```
Done

newtoncn commented on src/gate_detect/include/GateDetectionNode.h:
```diff
@@ -0,0 +1,63 @@
+     * Dynamic reconfigure callback function
+     *
+     * @param config
+     * @param level
```
Done

newtoncn commented on src/gate_detect/include/GateDetectionNode.h:
```diff
@@ -0,0 +1,63 @@
+#ifndef GATE_DETECT_GATEDETECT_H
+#define GATE_DETECT_GATEDETECT_H
+
+#include "../test/testUtilities/TestUtils.h"
```
Done

newtoncn commented on src/gate_detect/include/GateDetectionNode.h:
```diff
@@ -0,0 +1,63 @@
+
+    /**
+     * Publish image of gate seen by node, use for testing
+     * @param gateVector
```
Done

newtoncn commented on src/gate_detect/include/PoleDetection.h:
```diff
@@ -0,0 +1,52 @@
+#include <opencv2/imgproc/imgproc.hpp>
+
+class PoleDetection {
+    cv::Vec4i _side1, _side2;
```
Could be left and right or top and bottom

newtoncn commented on src/gate_detect/package.xml:
```diff
@@ -2,27 +2,71 @@
+  <export>
+    <!-- Other tools can request additional information be placed here -->
+
+  </export>
```
Done

newtoncn commented on src/gate_detect/include/PoleDetection.h:
```diff
@@ -0,0 +1,52 @@
+
+class PoleDetection {
+    cv::Vec4i _side1, _side2;
+    float _m, _b;
```
Done

newtoncn commented on src/gate_detect/src/GateDetection.cpp:
```diff
@@ -0,0 +1,330 @@
+    m_Hor = m_Vert = 81.88;
+
+    b_Hor = b_Vert = -.92791;
+}
```
They are rosparams, but they're reused in the dynamic reconfig callback so when the GateDetection node is initially instantiated it needs to have some default values.

newtoncn commented on src/gate_detect/src/GateDetection.cpp:
```diff
@@ -0,0 +1,330 @@
+    m_Hor = m_Vert = 81.88;
+
+    b_Hor = b_Vert = -.92791;
+}
```
The interpolation constants seem a bit sketchy to put as a rosparam because they should be set during calibration and not changed, but we could potentially do that

newtoncn commented on src/gate_detect/include/PoleDetection.h:
```diff
@@ -0,0 +1,52 @@
+#include <opencv2/highgui.hpp>
+#include <opencv2/imgproc/imgproc.hpp>
+
+class PoleDetection {
```
done

newtoncn commented:
Fixed the naming stuff

newtoncn commented:
@garethellis0 @reoliveira Should be good for review now, I handled most of the comments and any others that come up I can always adjust with a smaller PR

garethellis0 commented on src/gate_detect/src/GateDetector.cpp:
```diff
@@ -0,0 +1,362 @@
+
+    VertInterpolationConstant2_ = HorInterpolationConstant2_ = 81.88;
+
+    VertInterpolationConstant1_ = HorInterpolationConstant1_ = -.92791;
```
We're definitly going to need some way of setting these magic numbers (and they should probably be in a launch file).

newtoncn commented:
@garethellis0 needs to be merged by a non-contributing write-access holder
